### ISSUE TITLE
${#array} should trigger sc2128

### ISSUE THREAD

**Issue by izabera** [koalaman/shellcheck#780]
```sh
$ shellcheck -s bash - << 'eof'
a=(a b c); echo "${#a}"
a=(a b c); echo "$a"
eof


In - line 2:
a=(a b c); echo "$a"
                 ^-- SC2128: Expanding an array without an index only gives the first element.


```
no error on line 1 but it's the same problem

**Comment by jFransham** [comment]
The parser treats `"${#a}"` as braces containing `#a` and then tries to look that up as a variable name. This is also true of `"${a:-1}"`, for example. This would require adding a fairly complex bash extension parser. You could probably get away with having a shell-specific parse tree transformation function (but keep the tree data structure the same across shells).

This would probably look something like

```haskell
data Token =
    T_Binary Id String Token Token
    -- ...snip...
    -- If Bash/Fish/Zsh/whatever shells have different string length semantics
    -- then we can have seperate data for each
    | T_StringLen Id Token
    | T_Default Id Token Token
    -- plus whatever else
    -- ...snip...

-- Actual implementation will be more complex, obviously
parseBashExts :: Token -> Maybe Token
parseBashExts (T_DollarBraced iOuter (T_Literal iInner ('#':rest))) =
    Just . T_DollarBraced iOuter . T_Literal iInner . T_StringLen $ T_Literal whateverId rest
parseBashExts _ = Nothing
```
